### ISSUE TITLE
Display different runs side by side

### ISSUE THREAD

**Issue by dmasad** [projectmesa/mesa#573]
In Mesa (as in most ABM tools) the interactive visualization shows one model run at a time. Exploring different parameter values requires stopping the model, changing the value, and restarting. However, one way a model can illustrate the effect of a parameter is by showing multiple runs side by side, each with different values. Mesa could add a lot of value by making it easy to set up, run, and view side-by-side model instantiations (possibly with the same random seed) with different parameter values.

An example of what it could look like is here:
https://imgur.com/a/8M7q8

**Comment by Corvince** [comment]
I love the idea of this feature. I haven't seen this functionality in any other ABM framework, but I think it would be extremely useful. 

Can we make a roadmap for this?

The immediate challenge I see comes with the ModularServer. IIRC it is somewhat locked to a model class and model instance simultaneously. We would need to disentangle the model properties (shared among all model runs) and the communication with each model instance/run. 

I think the visualization should be more straight-forward. Shouldn't be a problem to display more than one visualization element and I can already imagine having linked and unlinked user-settable parameters, so you can change some parameters for all model runs and some per model run.

I would love to help with this feature, but I really wouldn't know where to start at the moment.

**Comment by Corvince** [comment]
After peaking into the code and making some minimal experiments, I have to revise my initial statement. It is rather trivial to run multiple models simultaneously server-side (albeit trivial only with the same parameters, but quite doable otherwise). 

But the visualization is more difficult. Maybe because I am not so used to javascript, but maybe also because its code is scattered around in so many files. Here is the result of a hard-coded experiment:

![multitest](https://user-images.githubusercontent.com/13568919/45804484-85d4f600-bcbc-11e8-8f75-26a823d9b686.png)

For the model I only had to change a couple of lines in `ModularVisualization.py`, basically changeing `self.model` to a list of two models. 

For the visualization I had to change  `ModularVisualization.py`, `CanvasGridVisualization.py`, `CanvasModule.js`, `runcontrol.js` and `modular_template.html` in tandem. But hey, I think the result is promising.

**Comment by Corvince** [comment]
I just couldn't leave my hands off of this (procrastination, yay). After some experimentation I have a solution that seems to work rather nice. 

![side_by_side](https://user-images.githubusercontent.com/13568919/46084089-20dd3c80-c1a3-11e8-91e7-fb48dce8bb7b.png)

Here is how I did it:
The ModularServer takes a new keyword `n_simulations` that tells it how many simulations to run in parallel. 
The individual Visualization modules (in Python) take _can_ include a parameter `n_sims'
The ModularServer creates a list of visualization elements as before and passes it to the Javascript modules

Now the individual Javascript Modules recieve this new `n_simulations` parameter and decide how they handle it: In the example above the CanvasModule decides to display grids side-by-side, while the (new) MultiChartModule plots all three simulations on the same chart. 


A big advantage of this implementation (although admittedly accidentally) is that the current VizModules continue to work as before (they just drop `n_simulations`. Although eventually we should upgrade all vizmodules, we could already deploy side-by-side runs with limited Viz support and then upgrade the vizmodules over time (and not have to rewrite ALL viz modules right now). For example, the "normal" ChartModule still displays only values for the first model. 

[Here is the PR](https://github.com/projectmesa/mesa/pull/585)

And the ToDo:
- [ ] Polish current implementation (e.g. un-gray MultiChart)
- [ ] Change User Params for simulations individually 
- [ ] Create convincing example for the added benefit.
- [ ] Upgrade all viz elements

**Comment by dmasad** [comment]
I'm finally getting around to playing with this. What you've done so far is extremely cool, and I'm excited for it to be implemented. A few initial observations:

- It looks like the FPS slider is being tripled up as well (or at least is off some other way), which is producing weird behaviors.

- The parallel models aren't instantiated with the same seed. This should probably be an optional parameter; in some cases, the user wants to be able to just run `N` models simultaneously, but in other cases it's more useful to launch all the models with identical initial conditions. 

- I wonder if it would be better to subclass` ModularServer` (and possibly create a parallel-vis specific HTML template) instead of trying to have one server class and front-end template be able to handle both one and multiple simulation runs.

I'll keep playing around with it, and try to work on some of those issues as well.

**Comment by Corvince** [comment]
> It looks like the FPS slider is being tripled up as well (or at least is off some other way), which is producing weird behaviors.

Yes, I guess thats because some of the CSS attributes, but didn't know how prevent this

> The parallel models aren't instantiated with the same seed. This should probably be an optional parameter; in some cases, the user wants to be able to just run N models simultaneously, but in other cases it's more useful to launch all the models with identical initial conditions.

Good point! That would be a nice option, although you could set up your model already in a way that it always uses the same seed. My current implementation just starts n model instances, and if they have a fixed seed it would always be the same. 

> I wonder if it would be better to subclass ModularServer (and possibly create a parallel-vis specific HTML template) instead of trying to have one server class and front-end template be able to handle both one and multiple simulation runs.

I think in the end most changes to the ModularServer are very straight-forward and everything works more or less the same. Right now there shouldn't even be any breaking changes. So its more about enabling multi-simulation support for the individual elements. I think branching out the multi-sim aspects would be much more demanding to keep up to date and in sync with changes to the single-sim part.

> (editing to add) A useful interface feature would be to have some sliders affect all the models, and some be model-by-model.

Yes I was thinking that the sliders would per default be collapsed into a single slider and affect all the simulations, but you can expand them to make simulation-specific changes. But so far I had no luck with the sliders anyway.


Having written all that, I am actually working on different approach to the visualisation with some cool features. I'll write something on the mailing list in the next few days about that. But regarding this issue I would suggest splitting it and first focusing on supporting multiple simulations server-side and then tackle the visualization later. But for the former we should further discuss or decide if we want to subclass the ModularServer or not.

**Comment by tpike3** [comment]
Closing to make issues more impactful and relevant to current Mesa issues (anything before Mesa 3.0 release Nov 9, 2024) per May 26, 2026 dev meeting.
